usr/lib/common/mech_aes.c

```diff
@@ -4214,3 +4214,126 @@ CK_RV ckm_aes_xts_crypt(STDLL_TokData_t *tokdata,
 
     return rc;
 }
+
+CK_RV aes_xts_cipher(CK_BYTE *in_data, CK_ULONG in_data_len,
```

Comment:
What was the reason to add this generic implementation? I mean, yes, it's good to have one, but was it not possible to use openssl with a bit of glue code for this?

Reply:
OpenSSL does not support multi-part operations with AES-XTS ! 

The OpenSSL man page states:

_The XTS implementation in OpenSSL does not support streaming. That is there must only be one EVP_EncryptUpdate call per EVP_EncryptInit_ex call (and similarly with the "Decrypt" functions)._

So with the generic implementation we can implement multi-part operations for the SoftToken by using OpenSSL's AES-ECB functions, also see commit message of Soft Token commit: https://github.com/opencryptoki/opencryptoki/pull/614/commits/a0a502ca6ec757cea9a7248219d086a956eb1ca9

Reply:
... and it also helps to implement AES-XTS with multi part support for the EP11 token (with protected keys) for example. The 2 callback functions corelate to what the PCC and KM instructions do....